cli/util/uploader.go

```diff
@@ -0,0 +1,432 @@
+	"github.com/urfave/cli/v2"
+)
+
+const (
+	searchBatchSize     = 10000       // Number of objects to search in a batch for finding max block in container.
+	maxParallelSearches = 40          // Control the number of concurrent searches.
+	oidSize             = sha256.Size // Size of object ID.
+)
+
+func uploadBin(ctx *cli.Context) error {
+	if err := cmdargs.EnsureNone(ctx); err != nil {
+		return err
+	}
+	rpcNeoFS := ctx.StringSlice("fs-rpc-endpoint")
+	containerIDStr := ctx.String("container")
+	attr := ctx.String("block-attribute")
+	indexFileAttribute := ctx.String("index-attribute")
+	acc, _, err := options.GetAccFromContext(ctx)
+	if err != nil {
+		return cli.Exit(fmt.Sprintf("failed to load account: %v", err), 1)
+	}
+
+	var containerID cid.ID
+	if err = containerID.DecodeString(containerIDStr); err != nil {
+		return cli.Exit(fmt.Sprintf("failed to decode container ID: %v", err), 1)
+	}
+
+	gctx, cancel := options.GetTimeoutContext(ctx)
+	defer cancel()
+	rpc, err := options.GetRPCClient(gctx, ctx)
+	if err != nil {
+		return cli.Exit(fmt.Sprintf("failed to create RPC client: %v", err), 1)
+	}
+
+	currentBlockHeight, err := rpc.GetBlockCount()
+	if err != nil {
+		return cli.Exit(fmt.Sprintf("failed to get current block height from RPC: %v", err), 1)
+	}
+	fmt.Fprintln(ctx.App.Writer, "Chain block height:", currentBlockHeight)
+
+	signer := user.NewAutoIDSignerRFC6979(acc.PrivateKey().PrivateKey)
+
+	p, err := pool.New(pool.NewFlatNodeParams(rpcNeoFS), signer, pool.DefaultOptions())
+	if err != nil {
+		return cli.Exit(fmt.Sprintf("failed to create NeoFS pool: %v", err), 1)
+	}
+
+	if err = p.Dial(context.Background()); err != nil {
+		return cli.Exit(fmt.Sprintf("failed to dial NeoFS pool: %v", err), 1)
+	}
+	defer p.Close()
+
+	maxBlockIndex, err := fetchMaxBlockIndex(ctx.Context, p, containerID, acc.PrivateKey(), attr)
+	if err != nil {
+		return cli.Exit(fmt.Sprintf("failed to fetch max block index from container: %v", err), 1)
+	}
+
+	fmt.Fprintln(ctx.App.Writer, "Blocks uploaded in NeoFS container:", maxBlockIndex)
+
+	if maxBlockIndex >= int(currentBlockHeight) {
+		fmt.Fprintln(ctx.App.Writer, "No new blocks to upload. Max index in NeoFS: ", maxBlockIndex, "current height:", currentBlockHeight)
+		return nil
+	}
+
+	const (
+		maxRetries     = 3                      // Maximum number of retries
+		initialBackoff = 500 * time.Millisecond // Initial backoff duration
+		backoffFactor  = 2                      // Backoff multiplier
+		maxBackoff     = 10 * time.Second       // Maximum backoff duration
+
+		numWorkers   = 10             // Number of concurrent block uploaders
+		numProducers = numWorkers * 2 // Number of concurrent block fetchers
+	)
+
+	blockChan := make(chan *block.Block, 20000)
+	errorChan := make(chan error, 1)
+	doneChan := make(chan struct{})
+
+	producerWg := sync.WaitGroup{}
+	producerWg.Add(numProducers)
+
+	// Retry function with exponential backoff
+	retry := func(action func() error) error {
```

Comment:
Should we save it? Maybe yes? I like it

Reply:
`2024/09/26 20:46:33 Failed to fetch block 413230: failed to fetch block 413230: Post "https://rpc.morph.fs.neo.org": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
`
`Upload error: failed to fetch block 219272: Post "https://rpc.t5.n3.nspcc.ru:20331": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
`
I think this one

also from history, but with slicer:
```
failed to upload block 150079: failed to create slicer: network info: status: code = 1024 message = websocket: close 1006 (abnormal closure): unexpected EOF
2024/09/26 21:28:26 Failed to upload block 630944: failed to create slicer: network info: write request: rpc error: code = Unavailable desc = unexpected HTTP status code received from server: 502 (Bad Gateway); transport: received unexpected content-type "text/html"

failed to fetch max block index from container: failed to search objects from 321129 to 421128: error during object IDs iteration: rpc error: code = Internal desc = unexpected EOF


```